File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.py
class Solution(object):
    def minFallingPathSum(self, mat):
        N = len(mat)
        for i in range(1, N):
            for j in range(N):
                mat[i][j] += min(mat[i - 1][k] for k in (j - 1, j, j + 1) if 0 <= k < N)
        return min(mat[-1])
        

File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.java
class Solution {
    public int minFallingPathSum(int[][] mat) {
        int n = mat.length;
        for(int i = 1; i<n; i++) {
            for(int j = 0; j<n; j++) {
                mat[i][j] += Math.min(mat[i-1][j], Math.min(mat[i-1][Math.max(0, j - 1)], mat[i-1][Math.min(n-1, j + 1)]));
            }
        }

        return Arrays.stream(mat[n-1]).min().getAsInt();
    }
}